### PR TITLE
executor: fix time filter condition when select from information.slow_query_log

### LINKED ISSUE #34320: select from information.slow_query_log get wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

cat tidb-slow.log

```
# Time: 2022-04-21T14:44:54.103041447+08:00
# Txn_start_ts: 432674816242745346
# Query_time: 59.251052432
# Parse_time: 0
# Compile_time: 21.36997765
# Rewrite_time: 2.107040149
# Optimize_time: 12.866449698
# Wait_TS: 1.485568827
# Cop_time: 8.619838386 Request_count: 1 Total_keys: 1 Rocksdb_block_cache_hit_count: 3
# Index_names: [bind_info:time_index]
# Is_internal: true
# Digest: caf0da652413a857b1ded77811703043e52753ca8a466e20e89c6b74d9662783
# Stats: bind_info:pseudo
# Num_cop_tasks: 1
# Cop_proc_avg: 0 Cop_proc_addr: 172.16.6.173:40161
# Cop_wait_avg: 0 Cop_wait_addr: 172.16.6.173:40161
# Mem_max: 186
# Prepared: false
# Plan_from_cache: false
# Plan_from_binding: false
# Has_more_results: false
# KV_total: 4.032247202
# PD_total: 0.108570401
# Backoff_total: 0
# Write_sql_response_total: 0
# Result_rows: 0
# Succ: true
# IsExplicitTxn: false
# Plan: tidb_decode_plan('8gW4MAkxNF81CTAJMzMzMy4zMwlteXNxbC5iaW5kX2luZm8udXBkYXRlX3RpbWUsIG06HQAMY3JlYQ0ddAkwCXRpbWU6MTkuM3MsIGxvb3BzOjEJMCBCeXRlcxEIIAoxCTMwXzEzCRlxFTkINy40GTkYLCAJMTg2IAk9OE4vQQoyCTQ3XzExCTFfMBWsFHRhYmxlOhWsHCwgaW5kZXg6AYgAXwULCCh1cBW+OCksIHJhbmdlOigwMDAwLQUDDCAwMDoFAwAuARSgMDAsK2luZl0sIGtlZXAgb3JkZXI6ZmFsc2UsIHN0YXRzOnBzZXVkbwkN6wg5LjYysgDAY29wX3Rhc2s6IHtudW06IDEsIG1heDogNS4wNnMsIHByb2Nfa2V5czogMCwgcnBjXxEmAQwBtRw6IDQuMDVzLAFKSHJfY2FjaGVfaGl0X3JhdGlvOiABphh9LCB0aWt2CWgAewU1ADA5Nlh9LCBzY2FuX2RldGFpbDoge3RvdGFsXwF6CGVzcxl9RhcAFF9zaXplOgGZCRwAawWogDEsIHJvY2tzZGI6IHtkZWxldGVfc2tpcHBlZF9jb3VudAUyCGtleUoWAAxibG9jIQsZxw0yFDMsIHJlYS5BAAUPCGJ5dAGBKfMYfX19CU4vQQEEIfoQNV8xMgly+gGCsgEgCU4vQQlOL0EK')
# Plan_digest: c338c3017eb2e4980cb49c8f804fea1fb7c1104aede2385f12909cdd376799b3
SELECT original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation, source  FROM mysql.bind_info WHERE update_time > '0000-00-00 00:00:00' ORDER BY update_time, create_time;
```

### 2. What did you expect to see? (Required)

```
mysql> select count(digest) from information_schema.SLOW_QUERY where time>'2022-04-29 17:50:00';
+---------------+
| count(digest) |
+---------------+
|            0 |
+---------------+
1 row in set (0.04 sec)
```

### 3. What did you see instead (Required)

```
mysql> select count(digest) from information_schema.SLOW_QUERY where time>'2022-04-29 17:50:00';
+---------------+
| count(digest) |
+---------------+
|            1 |
+---------------+
1 row in set (0.04 sec)
```

The `time>'2022-04-29 17:50:00'` filter condition doesn't work...
When I `select time, digest from information_schema.SLOW_QUERY where time>'2022-04-29 17:50:00';`  the result is correct.
After some testing, I found when the output column contains `time`, the time filter condition works, otherwise it scan the whole table logs.

This bug was introduced by https://github.com/pingcap/tidb/pull/26219/ since tidb 5.2.0
Before 5.1.0 it's OK.

### 4. What is your TiDB version? (Required)

master 6bd54bea8a9ec25c8d65fcf1157c5ee7a141ab0b

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #34320

Problem Summary:

### What is changed and how it works?

Such kind of requests can't get the correct result, the time filter condition doesnt work and the query scan the whole table. 

```
select count(digest) from information.slow_query where time > xxx
select digest, count(1) from information.slow_query group by digest order by count(1)
```

That's because when the slow query MemTable's output column doesn't contain 'time' column, it will not decode the 'time' column, thus ignore the 'time' filter condition.

When 'time' column is not in the output column, `e.columnValueFactoryMap[field]` is nil when `field` is `variable.SlowLogTimeStr`.
I add an extra check to decode the 'time' column and check the time filter condition in these cases.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/slow_query.go [hunk 1]
```diff
@@ -663,7 +663,20 @@ func (e *slowQueryRetriever) parseLog(ctx context.Context, sctx sessionctx.Conte
 func (e *slowQueryRetriever) setColumnValue(sctx sessionctx.Context, row []types.Datum, tz *time.Location, field, value string, checker *slowLogChecker, lineNum int) bool {
 	factory := e.columnValueFactoryMap[field]
 	if factory == nil {
-		return true
+		switch {
+		// Fix issue 34320, when slow log time is not in the output columns, the time filter condition is mistakenly discard.
+		case field == variable.SlowLogTimeStr && checker != nil:
```
Comment (crazycs520) [inline executor/slow_query.go:668]: The `user` column maybe has the same problem, neet also handle `user` column here.

Comment (tiancaiamao) [reply executor/slow_query.go:668]: No, the logic for `time` column and `user` column is different, the predication conditions on `time` column will be pushed down to the `MemTable`, while the predication conditions on the other columns use the `Selection` executor, so they don't have the save problem. @crazycs520

Comment (crazycs520) [reply executor/slow_query.go:668]: oh, yes

**Review by crazycs520 — APPROVED** [review]

File: executor/slow_query.go (hunk 1 above)
Comment (zimulala) [inline executor/slow_query.go:668]: Could we use `if` instead of `switch`? Here is only one `case`.

File: executor/slow_query.go [hunk 2]
```diff
@@ -663,7 +663,20 @@ func (e *slowQueryRetriever) parseLog(ctx context.Context, sctx sessionctx.Conte
...
+			t, err := ParseTime(value)
+			if err != nil {
+				err = fmt.Errorf("Parse slow log at line %v, failed field is %v, failed value is %v, error is %v", lineNum, field, value, err)
+				sctx.GetSessionVars().StmtCtx.AppendWarning(err)
+				return false
+			}
+			timeValue := types.NewTime(types.FromGoTime(t), mysql.TypeTimestamp, types.MaxFsp)
+			return checker.isTimeValid(timeValue)
```
Comment (zimulala) [inline executor/slow_query.go:676]: Do we need to check `checker != nil`? I see we do it in `getColumnValueFactoryByName`.
Besides, do we need to handle it with timezone？

File: executor/slow_query.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/slow_query.go:668]: Done

File: executor/slow_query.go (hunk 2 above)
Comment (tiancaiamao) [reply executor/slow_query.go:676]: `checker != nil` has already been check before entering this code branch. See https://github.com/pingcap/tidb/pull/34322/files#diff-9cb39ec0c8d4960b6e8f64d27731000897c8296503bad061a574b3dd8ac17c11R668

> do we need to handle it with timezone

You can see this special order https://github.com/pingcap/tidb/blob/ca900e6a8381bc7e55717c65e4031e5961c97092/executor/slow_query.go#L715-L725
I asked @crazycs520 is this a bug, should it be handle timezone first and `check()` then... he confirmed this is not bug.

So it means `check()` use the unhandled timezone...

Comment (zimulala) [reply executor/slow_query.go:675]: Got it.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**